projects/driver_controls/src/stm32f0xx/ads1015.c

```diff
@@ -0,0 +1,169 @@
+StatusCode ads1015_init(Ads1015Storage *storage, I2CPort i2c_port, Ads1015Address i2c_addr,
+                        GPIOAddress *ready_pin) {
+  if ((storage == NULL) || (ready_pin == NULL)) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  memset(storage, 0, sizeof(*storage));
+  for (uint8_t i = 0; i < NUM_ADS1015_CHANNELS; i++) {
+    storage->channel_readings[i] = ADS1015_DISABLED_CHANNEL_READING;
+  }
+  storage->i2c_port = i2c_port;
+  storage->i2c_addr = i2c_addr + ADS1015_I2C_BASE_ADDRESS;
+  storage->ready_pin = *ready_pin;
+  // Set up config register.
+  status_ok_or_return(prv_setup_register(storage, ADS1015_ADDRESS_POINTER_CONFIG,
+                                         ADS1015_CONFIG_REGISTER_MSB(ADS1015_CHANNEL_0),
+                                         ADS1015_CONFIG_REGISTER_LSB));
+  // Set up hi/lo-thresh registers. This particular setup enables the ALRT/RDY pin.
+  status_ok_or_return(prv_setup_register(storage, ADS1015_ADDRESS_POINTER_LO_THRESH,
+                                         ADS1015_LO_THRESH_REGISTER_MSB,
+                                         ADS1015_LO_THRESH_REGISTER_LSB));
+  status_ok_or_return(prv_setup_register(storage, ADS1015_ADDRESS_POINTER_HI_THRESH,
+                                         ADS1015_HI_THRESH_REGISTER_MSB,
+                                         ADS1015_HI_THRESH_REGISTER_LSB));
+  GPIOSettings gpio_settings = {
+    .direction = GPIO_DIR_IN,  //
+  };
+  InterruptSettings it_settings = {
+    .type = INTERRUPT_TYPE_INTERRUPT,
+    .priority = INTERRUPT_PRIORITY_NORMAL,
+  };
+  status_ok_or_return(gpio_init_pin(ready_pin, &gpio_settings));
+  status_ok_or_return(gpio_it_register_interrupt(ready_pin, &it_settings, INTERRUPT_EDGE_RISING,
+                                                 prv_interrupt_handler, storage));
+  // Mask the interrupt until channels are enabled by the user.
+  return gpio_it_mask_interrupt(ready_pin, true);
+}
+
+// This function enable/disables channels, and registers callbacks for each channel.
+StatusCode ads1015_configure_channel(Ads1015Storage *storage, Ads1015Channel channel, bool enable,
+                                     Ads1015Callback callback, void *context) {
+  if (storage == NULL || channel >= NUM_ADS1015_CHANNELS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  if (storage->channel_enable_bitset == ADS1015_BITSET_EMPTY && enable) {
+    // Activate the interrupt since the first channel is being enabled.
+    status_ok_or_return(gpio_it_mask_interrupt(&storage->ready_pin, false));
+    status_ok_or_return(prv_set_channel(storage, channel));
+  } else if (!enable) {
+    storage->channel_readings[channel] = ADS1015_DISABLED_CHANNEL_READING;
+  }
+
+  prv_mark_channel_enabled(channel, enable, &storage->channel_enable_bitset);
+  if (storage->channel_enable_bitset != ADS1015_BITSET_EMPTY) {
+    storage->current_channel = __builtin_ffs(storage->channel_enable_bitset) - 1;
```

Comment:
Doesn't this end up changing the bitset every time a channel is configured?

i.e. if you're already on channel 2 and you enable channel 0, you'll set the channel to channel 0.

Reply:
Yah but I did that too prevent staying on a channel that is being disabled. That's actually how I fixed the bug. Channel0 was being disabled but the current channel was not changing to another channel. Also changing the channel just once in config does not seem to have a bad outcome.

Reply:
But you're introducing a new bug - If you previously started a conversion for channel 2, then you set the current channel to 0, when your alert interrupt runs, you'll read back the data for channel 2 but store it in channel 0.